library/Booster/Pattern/Rewrite.hs

```diff
@@ -320,24 +320,31 @@ applyRule pat@Pattern{ceilConditions} rule = runRewriteRuleAppT $ do
             Just False -> RewriteRuleAppT $ pure Trivial
             _other -> pure ()
 
+    -- existential variables may be present in rule.rhs and rule.ensures,
+    -- need to strip prefixes and freshen their names with respect to variables already
+    -- present in the substitution
+    let varsFromSubst = Set.unions . map freeVariables . Map.elems $ subst
+        existentialSubst =
+            Map.fromSet
+                (\v -> Var $ freshenVar v{variableName = stripVarOriginPrefix v.variableName} varsFromSubst)
+                rule.existentials
+
+    -- modify the substitution to include the existentials
+    let substWithExistentials = subst `Map.union` existentialSubst
+
     let rewritten =
             Pattern
-                (substituteInTerm (refreshExistentials subst) rule.rhs)
+                (substituteInTerm substWithExistentials rule.rhs)
                 -- adding new constraints that have not been trivially `Top`
-                ( Set.fromList newConstraints
-                    <> Set.map (coerce . substituteInTerm subst . coerce) pat.constraints
+                ( Set.map
+                    (coerce . substituteInTerm substWithExistentials . coerce)
+                    (pat.constraints <> Set.fromList newConstraints)
                 )
-                ceilConditions
+                (map (coerce . substituteInTerm substWithExistentials . coerce) ceilConditions)
```

Comment:
Existentials won't end-up in ceils, but it does not hurt to apply the modified substitution.

Reply:
On second thought, I've decided to revert that and not to touch the ceils.